### PR TITLE
Copy `cortex/pkg/cortexpb` package dependency into Loki

### PR DESCRIPTION
**Pull request by DylanGuedes**
**What this PR does / why we need it**:
As a part of removing Loki dependency on cortex packages, we decided to copy the  `cortexpb` package from cortex to Loki itself.
A few things:
- **Sample `TimestampMs` is now `Timestamp`, because the existing `logproto` has it named this way**. I'm not sure if this is a problem or not, but at the least the tests are all passing
- After some internal discussion, I decided for merging it with the existing `logproto` package, so I had to rename a few things
- We can't yet get rid of it as some vendored packages still relying on it. Once we remove those we can entirely remove `cortexpb`.

**Which issue(s) this PR fixes**:
N/A

**Special notes for your reviewer**:
N/A

**Checklist**
- [ ] Documentation added
- [ ] Tests updated
- [ ] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

**Review by sandeepsukhani — COMMENTED** [review]
I can still see the usage of it in a couple of places, e.g https://github.com/grafana/loki/blob/bfd4a3db76a7ec06a9fb3d17c5c1d2c27e49d686/pkg/ruler/base/compat.go#L8
This could be the reason for dependency not getting dropped. Can you please check?

**Comment by DylanGuedes** [comment]
> I can still see the usage of it in a couple of places, e.g
> 
> https://github.com/grafana/loki/blob/bfd4a3db76a7ec06a9fb3d17c5c1d2c27e49d686/pkg/ruler/base/compat.go#L8
> 
> 
> This could be the reason for dependency not getting dropped. Can you please check?

Great catch, thank you :tada: 

For the other cases (ex: in the `querier` package), unfortunately we can't yet move away from `cortexpb` as they are still using Cortex `ingester/client`, which expects the `cortexpb` implementation. Moreover, even after all those replaces, we won't be able to get get rid of it for the reason listed in the PR description:

> We can't yet get rid of it as some vendored packages still relying on it. Once we remove those we can entirely remove cortexpb.

To completely get rid of `cortexpb` we have to:
- Fork `ingester/client`
- Fork `chunk/purger`
- Fork `util/validation`
- Fork `util/limiter`
- Fork `util/extract`

The `ingester/client` fork is already opened and the others seems easy, so I'm confident we'll remove it this week :pray:

File: pkg/logproto/compat.go [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+
+	for i, s := range samples {
+		ts := TimeseriesFromPool()
+		ts.Labels = append(ts.Labels, FromLabelsToLabelAdapters(lbls[i])...)
+		ts.Samples = append(ts.Samples, s)
+		req.Timeseries = append(req.Timeseries, PreallocTimeseries{TimeSeries: ts})
+	}
+
+	return req
+}
```
Comment (cyriltovena) [inline pkg/logproto/compat.go:39]:
```suggestion

```

File: pkg/logproto/compat.go [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+	result := make([]exemplar.Exemplar, 0, len(es))
+	for _, e := range es {
+		result = append(result, exemplar.Exemplar{
+			Labels: FromLabelAdaptersToLabels(e.Labels),
+			Value:  e.Value,
+			Ts:     e.TimestampMs,
+		})
+	}
+	return result
+}
```
Comment (cyriltovena) [inline pkg/logproto/compat.go:142]:
```suggestion

```

File: pkg/logproto/compat.go [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+	case SUMMARY:
+		return textparse.MetricTypeSummary
+	case INFO:
+		return textparse.MetricTypeInfo
+	case STATESET:
+		return textparse.MetricTypeStateset
+	default:
+		return textparse.MetricTypeUnknown
+	}
+}
```
Comment (cyriltovena) [inline pkg/logproto/compat.go:173]:
```suggestion
```

File: pkg/logproto/metrics.proto [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+message Metric {
+  repeated LabelPair labels = 1 [(gogoproto.nullable) = false, (gogoproto.customtype) = "LabelAdapter"];
+}
+
+message Exemplar {
+  // Exemplar labels, different than series labels
+  repeated LabelPair labels = 1 [(gogoproto.nullable) = false, (gogoproto.customtype) = "LabelAdapter"];
+  double value = 2;
+  int64 timestamp_ms = 3;
+}
```
Comment (cyriltovena) [inline pkg/logproto/metrics.proto:61]:
```suggestion
```

**Review by cyriltovena — APPROVED** [review]
LGTM we might be able to remove some more.

File: pkg/logproto/compat.go (hunk 1 above)
Comment (DylanGuedes) [reply pkg/logproto/compat.go:39]: Hmmm I think this is used indirectly:

- `ToWriteRequest` is invoked by `PusherAppender.Commit()`.
  -  `PusherAppender` is instantiated during `PusherAppendable.Appender()`
    - `PusherAppendable` is instantiated under `defaultTenantManagerFactory`
      - `defaultTenantManagerFactory` is invoked by `buildRuler()`
        - `buildRuler()` is used by many tests under `ruler/base`

Do you have a suggestion on how to get rid of it?

File: pkg/logproto/compat.go (hunk 3 above)
Comment (DylanGuedes) [reply pkg/logproto/compat.go]: good catch, ty

File: pkg/logproto/metrics.proto (hunk 4 above)
Comment (DylanGuedes) [reply pkg/logproto/metrics.proto]: done! good catch :tada:

**Review by sandeepsukhani — APPROVED** [review]
LGTM

**Pull request merged**